source/postprocess/sea_level.cc

```diff
@@ -0,0 +1,584 @@
+  MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+  GNU General Public License for more details.
+
+  You should have received a copy of the GNU General Public License
+  along with ASPECT; see the file LICENSE.  If not see
+  <http://www.gnu.org/licenses/>.
+*/
+
+
+#include <aspect/simulator.h>
+#include <aspect/global.h>
+#include <aspect/mesh_deformation/free_surface.h>
+#include <aspect/utilities.h>
+#include <aspect/structured_data.h>
+#include <aspect/geometry_model/spherical_shell.h>
+#include <aspect/postprocess/sea_level.h>
+#include <aspect/postprocess/geoid.h>
+
+#include <deal.II/base/quadrature_lib.h>
+#include <deal.II/fe/fe_values.h>
+
+#include <cmath>
+#include <limits>
+
+// TODO:
+// Include geoid anomaly of surface load
+// Compute new surface gravity per location, dependent on surface topography and Earth model
+// Time-changing ocean geometry
+
+namespace aspect
+{
+  namespace Postprocess
+  {
+    template <int dim>
+    void
+    SeaLevel<dim>::initialize()
+    {
+      Assert(false, ExcNotImplemented());
+    }
+
+    template <>
+    void
+    SeaLevel<3>::initialize()
+    {
+      const int dim = 3;
+
+      topography_lookup = std::make_unique<Utilities::StructuredDataLookup<dim-1>>(1,1.0);
+      topography_lookup->load_file(data_directory_topography+data_file_name_topography,this->get_mpi_communicator());
+
+      ice_height_lookup = std::make_unique<Utilities::StructuredDataLookup<dim-1>>(1,1.0);
+      ice_height_lookup->load_file(data_directory_ice_height+data_file_name_ice_height,this->get_mpi_communicator());
+    }
+
+    template <int dim>
+    double
+    SeaLevel<dim>::compute_nonuniform_sea_level_change(const Point<dim> &/*position*/) const
+    {
+      Assert(false, ExcNotImplemented());
+      return 0;
+    }
+
+    template <>
+    double
+    SeaLevel<3>::compute_nonuniform_sea_level_change(const Point<3> &position) const
+    {
+      const int dim = 3;
+
+      const Postprocess::Geoid<dim> &geoid =
+        this->get_postprocess_manager().template get_matching_postprocessor<Postprocess::Geoid<dim>>();
+
+      const double geoid_displacement = geoid.evaluate(position); // (check sign of geoid_displacement)
+      const double topography = this->get_geometry_model().height_above_reference_surface(position);
+
+      Point<dim-1> internal_position;
+      const std::array<double,dim> spherical_position = aspect::Utilities::Coordinates::cartesian_to_spherical_coordinates(position);
+      for (unsigned int d = 1; d < dim; ++d)
+        internal_position[d-1] = spherical_position[d];
+
+      const double topography_init = topography_lookup->get_data(internal_position,0);
+      int ocean_mask = 0;
+      if (topography_init > 0.0)
+        ocean_mask = 1;
+      else
+        ocean_mask = 0;
+
+      const double nonuniform_sea_level_change = (geoid_displacement-topography)*ocean_mask;
+
+      return nonuniform_sea_level_change;
```

Comment:
I understand this is shorter and does the same. But I find the initial implementation easier to read/understand, because it's more descriptive, and easier to identify which part of the sea level equation this is.

Reply:
OK, that's fine then.